### PR TITLE
iox-#1129 non-static runtime/node

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR adds an experimental feature to be able to use the runtime as a non-static variant in a `Node`. The `Node` acts as factory to create publisher and subscriber. This is the preparatory work for running several RouDi in parallel.

The feature is intentionally not documented in a prominent way to prevent many user relying on it before it is ready to be enabled by default.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1129

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -60,6 +60,20 @@ Please make sure you have:
 
 **NOTE:** For support while developing you can use little helper scripts, see [git-hooks](./tools/git-hooks/Readme.md).
 
+## Experimental features
+
+Large features or features where the API is not yet clear can be implemented with the `IOX_EXPERIMENTAL` feature flag. Those
+features shall not be available in the installed headers when the `IOX_EXPERIMENTAL` feature flag was not set during compilation.
+
+If possible, this should be achieved not installing the headers of the experimental feature instead of `ifdefs`. With this
```
Comment (elfenpiff) [inline CONTRIBUTING.md:68]: *by not installing

File: iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+constexpr posh::experimental::NodeBuilderError
+from<runtime::IpcRuntimeInterfaceError, posh::experimental::NodeBuilderError>(
+    runtime::IpcRuntimeInterfaceError e) noexcept
+{
+    using namespace posh::experimental;
+    using namespace runtime;
+    switch (e)
+    {
+    case IpcRuntimeInterfaceError::CANNOT_CREATE_APPLICATION_CHANNEL:
+        return NodeBuilderError::IPC_CHANNEL_CREATION_FAILED;
```
Comment (elfenpiff) [inline iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp:110]: What do you think about adding some debug log nessages here, for the users and our future selves - when we receive bug reports.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 3]
```diff
@@ -0,0 +1,410 @@
...
+
+    auto node_result = RouDiEnvNodeBuilder("foo").roudi_registration_timeout(wait_for_roudi_timeout).create();
+
+    EXPECT_TRUE(timer.hasExpired());
+
+    ASSERT_TRUE(node_result.has_error());
+    EXPECT_THAT(node_result.error(), Eq(NodeBuilderError::TIMEOUT));
+}
+
+TEST(Node_test, RegisteringNodeWithDelayedRouDiStartWorks)
```
Comment (elfenpiff) [inline iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:160]: Please also test to create the maximum supported amount of nodes.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 4]
```diff
@@ -0,0 +1,410 @@
...
+
+    auto subscriber_result = node.subscriber({"all", "glory", "hypnotoad"}).create();
+    ASSERT_FALSE(subscriber_result.has_error());
+
+    auto subscriber = std::move(subscriber_result.value());
+
+    EXPECT_TRUE((std::is_same_v<decltype(subscriber), iox::unique_ptr<iox::posh::experimental::UntypedSubscriber>>));
+}
+
+TEST(Node_test, ExhaustingSubscriberLeadsToError)
```
Comment (elfenpiff) [inline iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:297]: Can you also add a test that exhausts Publishers, Subscribers and Waitsets

**Review by elfenpiff — APPROVED** [review]

File: iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp:110]: This is a `constexpr` function and also only translates error codes. I would keep it as is.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp (hunk 3 above)
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:160]: See https://github.com/eclipse-iceoryx/iceoryx/pull/2208/files#r1506312756

**Review by elfenpiff — APPROVED** [review]
